src/commons/authentication/oidc.rs

```diff
@@ -0,0 +1,366 @@
+    pub fn client_credentials_env_names(secret_name: &str) -> (String, String) {
+        let mut hasher = DefaultHasher::new();
+        secret_name.hash(&mut hasher);
+        let secret_name_hash = hasher.finish();
+
+        // Prefix with zeros to have consistent length. Max length is 16 characters, which is caused by [`u64::MAX`].
+        let secret_name_hash = format!("{:016x}", secret_name_hash).to_uppercase();
+        let env_var_prefix = format!("OIDC_{secret_name_hash}");
+
+        (
+            format!("{env_var_prefix}_CLIENT_ID"),
+            format!("{env_var_prefix}_CLIENT_SECRET"),
+        )
+    }
+
+    pub fn client_credentials_env_var_mounts(secret_name: String) -> Vec<EnvVar> {
+        let (client_id_env_var, client_secret_env_var) =
+            Self::client_credentials_env_names(&secret_name);
+
+        vec![
+            EnvVar {
+                name: client_id_env_var,
+                value_from: Some(EnvVarSource {
+                    secret_key_ref: Some(SecretKeySelector {
+                        key: CLIENT_ID_SECRET_KEY.to_string(),
+                        name: Some(secret_name.clone()),
+                        ..Default::default()
+                    }),
+                    ..Default::default()
+                }),
+                ..Default::default()
+            },
+            EnvVar {
+                name: client_secret_env_var,
+                value_from: Some(EnvVarSource {
+                    secret_key_ref: Some(SecretKeySelector {
+                        key: CLIENT_SECRET_SECRET_KEY.to_string(),
+                        name: Some(secret_name),
+                        ..Default::default()
+                    }),
+                    ..Default::default()
+                }),
+                ..Default::default()
+            },
+        ]
+    }
+}
+
+/// An enum of supported OIDC or identity providers which can serve as a hint
+/// in the product operator. Some products require special handling of
+/// authentication related config options. This hint can be used to enable such
+/// special handling.
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "PascalCase")]
+pub enum IdentityProviderHint {
+    Keycloak,
+}
+
+/// OIDC specific config options. These are set on the product config level.
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ClientAuthenticationOptions<T = ()> {
+    /// A reference to the OIDC client credentials secret. The secret contains
+    /// the client id and secret.
+    #[serde(rename = "clientCredentialsSecret")]
+    pub client_credentials_secret_ref: String,
+
+    /// An optional list of extra scopes which get merged with the scopes
+    /// defined in the [`AuthenticationClass`].
+    #[serde(default)]
+    pub extra_scopes: Vec<String>,
+
+    #[serde(flatten)]
+    pub extra_fields_for_product: T,
```

Comment:
This might be a little wordy. Yes, the intention is to make this very clear to the developer. We might want to shorten it to `extra_fields`, and add doc comments instead. If we decide to keep the name as is, we still would want doc comments on that struct field.

Reply:
As discussed: 151dcc6